### ISSUE TITLE
refmt issues with chained |>

### ISSUE THREAD

**Issue by kyldvs** [facebook/reason#1045]
When chaining `|>` refmt makes some weird decisions that I believe to be undesirable formatting. I start with:

```
let code = JSCodegen.Code.(
  create
  |> lines Requires.(
    create
    |> import_type local::"Set" source::"Set"
    |> import_type local::"Map" source::"Map"
    |> import_type local::"Immutable" source::"immutable"
    |> require local::"invariant" source::"invariant"
    |> require local::"Image" source::"Image.react"
    |> side_effect source::"monkey_patches"
    |> render_lines
  )
  |> new_line
  |> new_line
  |> new_line
  |> new_line
  |> render
);
```

And after formatting end with:

```
let code =
  JSCodegen.Code.(
    create |>
    lines
      Requires.(
        create |> import_type local::"Set" source::"Set" |>
        import_type local::"Map" source::"Map" |>
        import_type local::"Immutable" source::"immutable" |>
        require local::"invariant" source::"invariant" |>
        require local::"Image" source::"Image.react" |>
        side_effect source::"monkey_patches" |> render_lines
      ) |> new_line |> new_line |> new_line |> new_line |> render
  );

```

I believe there are 3 distinct issues with this new format:

1. There shouldn't need to be a break between `let code =` and `JSCodegen.Code.(`. This just makes all the code below it more likely to break lines and wastes horizontal space. I believe this also causes the break between `|> lines` and `Requires.(`, but that could be a separate 4th issue as well. (This appears to have been an unintended regression in https://github.com/facebook/reason/pull/812)
2. Once a single `|>` breaks in a chain, I believe they should all be on a separate line. For example there should be a break after `create |>` or more noticeably after each `new _line |>`
3. I think `|>` should start the line rather than end it. Admittedly I could see this being more of a preference thing, but being at the start feels more consistent with this kind of formatting:

```
type foo =
  | Bar string
  | Baz string
  | Buz string;
```

**Comment by kyldvs** [comment]
More support for (3) in the Reason source code itself :) (although .ml not a .re file...)

https://github.com/facebook/reason/blob/master/src/reason_format_type.ml#L24-L36

**Comment by kyldvs** [comment]
Here's my attempt at making some progress on this, probably a good place to start from (or maybe not, I have no idea :P)

https://github.com/facebook/reason/compare/master...kyldvs:pipe-chaining?expand=1

This:

```
let code = JSCodegen.Code.(
  create
  |> lines Requires.(
    create
    |> import_type local::"Set" source::"Set"
    |> import_type local::"Map" source::"Map"
    |> import_type local::"Immutable" source::"immutable"
    |> require local::"invariant" source::"invariant"
    |> require local::"Image" source::"Image.react"
    |> side_effect source::"monkey_patches"
    |> render_lines
  )
  |> new_line
  |> new_line
  |> new_line
  |> new_line
  |> render
);

let code = JSCodegen.Code.(create |> render);
```

Becomes:

```
let code =
  JSCodegen.Code.(
    create
    |> lines
         Requires.(
           create
           |> import_type local::"Set" source::"Set"
           |> import_type local::"Map" source::"Map"
           |> import_type local::"Immutable" source::"immutable"
           |> require local::"invariant" source::"invariant"
           |> require local::"Image" source::"Image.react"
           |> side_effect source::"monkey_patches"
           |> render_lines
         )
    |> new_line
    |> new_line
    |> new_line
    |> new_line
    |> render
  );

let code = JSCodegen.Code.(create
                           |> render);
```

**Comment by yunxing** [comment]
Good point. I think we will have more use cases like this (monadic binding like ">="). 

For the change, I'd keep the nodes as a list (instead of label) but try use the ["separators_stick_left"](https://github.com/facebook/reason/blob/master/vendor/easy_format/easy_format.mli#L86) option from easy_format.

**Comment by jordwalke** [comment]
I think `"separators_stick_left"` might not work with the way we do comment interleaving (just a thing to test, I can't recall for certain). (We reimplemented separators so that they aren't attached to the end of interleaved comments)

**Comment by yunxing** [comment]
@jordwalke we can have a check for this option to decide if we want to use the "reimplemented separators" logic. But I imagine this issue go away as we gradually hack into the easyformat's API.

**Comment by jordwalke** [comment]
I definitely agree with `#2` and I believe that as @yunxing mentioned "separators_stick_left" would also work here (and we should fix that feature that we removed due to comment interleaving).

But there could be other ways that are a little easier as a quick fix, such as ensuring they are a list instead of sequence of labels. That might *also* be difficult with the way the code is structured, so I would first try setting the `label` break policy for each of the infix operators to be `Always_rec`. That might be a very easy fix.

**Comment by chenglou** [comment]
Bumping this. Any news? It's one of the glaring flaws with refmt.

**Comment by kyldvs** [comment]
I haven't had time (and probably won't anytime soon) to work on my solution any further. For now I've just been using a local build of refmt with my hack in it

**Comment by chenglou** [comment]
Fixed by #1259
